### PR TITLE
Admin UI: custom success toast UI hook

### LINKED ISSUE #677: Specify custom "Success" messages per list
## What

Imagine this flow:

- I run a blog site where people can submit their own posts.
- I receive an email telling me I got a new post.
- I click the link to load up Keystone on that post.
- I now have to change the `status` to `Approved`
- I click `Save`
- _An email is sent to the author, **and** the post is now live on the site_
- I see:
  screen shot 2019-02-09 at 2 18 40 pm

That success message _"<label> Saved Successfully"_ is not very informative of what actually happened.

Ideally, I'd be able to specify a custom message to better describe what happened (an email was sent, and the post is live at `<url>`). Something like:

screen shot 2019-02-09 at 2 38 20 pm

## How

One possible solutions is as part of the `afterChange()` hook (which currently expects `undefined` as the return type);

```javascript
keystone.createList('Post', {
  fields: {
    approved: { type: Checkbox },
    /* ... */
  },
  hooks: {
    afterChange({ updatedItem, existingItem }) {
      if (updatedItem.approved && !existingItem.approved) {
        sendApprovalEmail(updatedItem);
        return {
          message: `
**${updatedItem.title}**

✅ Changes saved successfully.<br />
✉️ ${updatedItem.author.name} has been notified of approval.<br />
🔗 <a href="${updatedItem.url}">View the post online here</a>
          `,
        },
      }
    },
  },
});
```

### PR DESCRIPTION
**Pull request by Vultraz**
Built on top of #2816 since those changes allow me to pass the item data to the custom hook without exposing sensitive data like raw passwords. Relevant changes are here: https://github.com/keystonejs/keystone/commit/e144afcbf7c70aa7d2c4da30d7149f6c761133ff

Resolves #677. This add a new `toast${listKey}Saved` UI hook that allows users to customize Success toast content on a per-list basis. So for example, `toastUserSaved` would be used when saving the `User` list. Unlike the other hooks, this one receives the newly-saved item data *after* the refetch. I think everything should be sanitized at that point.

@jesstelford 

Example in action:
```js
  toastUserSaved: itemData => (
    <div>
      <div>Some custom toast content!</div>✅✅✅<span>{itemData._label_}</span><span>Stuff stuff stuff</span>
    </div>
  ),
};
```
image

### CODE REVIEW

**Comment by gautamsi** [comment]
the only problem I se is that hooks specification is not finalized which keeps several of my PR stale for long. Unless I do not have context of recent changes.

I had several chats and comments and i would like list specific things to be together with list config but seems like everything is going in one big hooks file ( I can not opposed to that).

Having flat structure which uses hook name combined with hook name like `toast[User]Saved` may not scale well. To me hooks file readability is hampered. 

We first need to :
*  finalize spec (naming convention, hierarchy levels and flat structure)
* List possible hooks
* guideline for naming new hooks out of possible hooks

I think this has been pending for long @MadeByMike @timleslie.

**Comment by Vultraz** [comment]
I'm definitely open to changing the implementation based on what people think. This is just one way this could be done.

File: packages/app-admin-ui/README.md [hunk 1]
```diff
@@ -122,6 +123,23 @@ export default {
...
 ```
 
+#### `toast${listKey}Saved`
+
+Allows customizing the content of the Success toast notification when an item is saved. These may be specified on a per-list basis. For example, to customize the Success toast for a list called `MyList`, use `toastMyListSaved`.
+
+Unlike most hooks, this takes a single argument containing the newly-saved item data. It should return a React component.
+
+```javascript
+export default {
```
Comment (MadeByMike) [inline packages/app-admin-ui/README.md:133]: I think we need a full example of how to use this with `require.resolve`

**Comment by Vultraz** [comment]
There isn't really a `Toast` component right now, though. I could definitely add one, but currently the layout data is just an argument to `addToast`.

**Comment by Vultraz** [comment]
Fixed conflicts.

**Comment by Vultraz** [comment]
Yeah, I'll defer to @MadeByMike on this.

**Comment by MadeByMike** [comment]
We're going for flat keys.

**Pull request merged**